tls/extensions/s2n_client_psk.c

```diff
@@ -157,21 +161,79 @@ static S2N_RESULT s2n_match_psk_identity(struct s2n_array *known_psks, const str
     return S2N_RESULT_OK;
 }
 
+int s2n_choose_psk(struct s2n_connection *conn, 
+                   struct s2n_psk_identity *identities, size_t identities_length,
+                   struct s2n_chosen_psk *chosen_psk)
+{
+    notnull_check(conn);
+    notnull_check(identities);
+    notnull_check(chosen_psk);
+
+    struct s2n_array *known_psks = &conn->psk_params.psk_list;
+
+    for (size_t i = 0; i < identities_length; i++) {
+        struct s2n_blob wire_identity = { 0 };
+        GUARD(s2n_blob_init(&wire_identity, identities[ i ].data, identities[ i ].length));
+
+        struct s2n_psk *local_match = NULL;
+        GUARD_AS_POSIX(s2n_match_psk_identity(known_psks, &wire_identity, &local_match));
+
+        if (local_match != NULL) {
+            chosen_psk->external_psk.identity_length = local_match->identity.size;
+            memcpy_check(chosen_psk->external_psk.identity, local_match->identity.data, local_match->identity.size);
+            chosen_psk->external_psk.secret_length = local_match->secret.size;
+            memcpy_check(chosen_psk->external_psk.secret, local_match->secret.data, local_match->secret.size);
+            GUARD(s2n_external_psk_set_hmac(&chosen_psk->external_psk, local_match->hmac_alg));
+            chosen_psk->wire_index = i;
+            break;
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_count_psk_identities(struct s2n_stuffer *input, uint16_t *identity_count)
+{
+    ENSURE_REF(input);
+
+    *identity_count = 0;
+    while (s2n_stuffer_data_available(input) > 0) {
+        uint16_t identity_size = 0;
+        GUARD_AS_RESULT(s2n_stuffer_read_uint16(input, &identity_size));
+        GUARD_AS_RESULT(s2n_stuffer_skip_read(input, identity_size));
+        GUARD_AS_RESULT(s2n_stuffer_skip_read(input, sizeof(uint32_t)));
+        (*identity_count)++;
+    }
+    GUARD_AS_RESULT(s2n_stuffer_reread(input));
+    return S2N_RESULT_OK;
+}
+
 static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn, struct s2n_stuffer *wire_identities_in)
 {
     ENSURE_REF(conn);
     ENSURE_REF(wire_identities_in);
+    ENSURE_EQ(conn->psk_params.chosen_psk, NULL);
+    ENSURE_EQ(conn->psk_params.chosen_psk_wire_index, 0);
+
+    uint16_t identities_count = 0;
+    GUARD_RESULT(s2n_count_psk_identities(wire_identities_in, &identities_count));
+
+    if (identities_count == 0) { return S2N_RESULT_OK; }
+
+    DEFER_CLEANUP(struct s2n_blob wire_identities_blob = { 0 }, s2n_free);
+    GUARD_AS_RESULT(s2n_alloc(&wire_identities_blob, identities_count * sizeof(struct s2n_psk_identity)));
+    struct s2n_psk_identity *wire_identities = ( struct s2n_psk_identity * )wire_identities_blob.data;
```

Comment:
Nit: your spacing is wonky here. `(struct s2n_psk_identity*) wire_identities_blob.data;`